pandas/core/indexes/period.py

```diff
@@ -148,144 +177,176 @@ class PeriodIndex(PeriodArrayMixin, DatelikeOps, DatetimeIndexOpsMixin,
+                # Some people seem to be calling use with unexpected types
+                # Index.difference -> ndarray[Period]
+                # DatetimelikeIndexOpsMixin.repeat -> ndarray[ordinal]
+                # I think that once all of Datetime* are EAs, we can simplify
+                # this quite a bit.
+                values = PeriodArray._complex_new(values, freq=self.freq)
+
+        # I don't like overloading shallow_copy with freq changes.
+        # See if it's used anywhere outside of test_resample_empty_dataframe
+        freq = kwargs.pop("freq", None)
+        if freq:
+            values = values.asfreq(freq)
+
+        attributes = self._get_attributes_dict()
+        attributes.update(kwargs)
+        if not len(values) and 'dtype' not in kwargs:
+            attributes['dtype'] = self.dtype
+        return self._simple_new(values, **attributes)
+
+    def _shallow_copy_with_infer(self, values=None, **kwargs):
         """ we always want to return a PeriodIndex """
         return self._shallow_copy(values=values, **kwargs)
 
-    def _coerce_scalar_to_index(self, item):
-        """
-        we need to coerce a scalar to a compat for our index type
+    # ------------------------------------------------------------------------
+    # Boxing
+    # err. maybe not...
 
-        Parameters
-        ----------
-        item : scalar item to coerce
+    @property
+    def _box_func(self):
+        def func(x):
+            if isinstance(x, Period) or x is tslib.NaT:
+                return x
+            else:
+                return Period._from_ordinal(ordinal=x, freq=self.freq)
+        return func
+
+    def _maybe_box_as_values(self, values, **attribs):
+        """Box an array of ordinals to a PeriodArray
+
+        This is purely for compatibility between PeriodIndex
+        and Datetime/TimedeltaIndex. Once these are all backed by
+        an ExtensionArray, this can be removed
         """
-        return PeriodIndex([item], **self._get_attributes_dict())
+        # TODO(DatetimeArray): remove
+        freq = attribs['freq']
+        return PeriodArray._from_ordinals(values, freq=freq)
+
+    # ------------------------------------------------------------------------
+    # Dispatch and maybe box. Not done in delegate_names because we box
+    # different from those (which use Index).
+
+    def asfreq(self, freq=None, how='E'):
+        result = self._data.asfreq(freq=freq, how=how)
+        return self._simple_new(result, name=self.name)
+
+    def _nat_new(self, box=True):
+        # TODO(DatetimeArray): remove this
+        result = self._data._nat_new(box=box)
+        if box:
+            result = self._simple_new(result, name=self.name)
+        return result
+
+    def to_timestamp(self, freq=None, how='start'):
+        from pandas import DatetimeIndex
+        result = self._data.to_timestamp(freq=freq, how=how)
+        return DatetimeIndex._simple_new(result,
```

Comment:
On principle shouldn't we be avoiding DatetimeIndex._simple_new here?  The performance impact can't be that big a deal can it?

Sidenote: probably should have a to_index method for DatetimeArray etc, so this last line just becomes `return result.to_index(name=self.name)`

Reply:
> On principle shouldn't we be avoiding DatetimeIndex._simple_new here?

In principle yes, but since our default constructors are that convoluted (and that is not something we can easily change), I think we should consider the `_simple_new` as an "internal public" method, meaning that we can use it throughout the pandas codebase (so outside it's own class definition where one can do `self._simple_new`)

> probably should have a to_index method for DatetimeArray etc

Personally, I think Arrays should be completely ignorant of the Index concept. That keeps a clearer separation of concerns

Reply:
> Personally, I think Arrays should be completely ignorant of the Index concept

That's fair.  `arr.to_index()` really isn't any less verbose than `pd.Index(arr)`